### PR TITLE
Restructure Photo module

### PR DESCRIPTION
**Pull request by jszanto**
This PR doe not change any functionality, however is does address a few small bugs. The main goal is to make some of the code more robust and easier to understand.

Changelog
-Album and photo counts are no longer cached in the database.
-Entity change event listeners now have their own classes.
-Upon removing a Photo or Album entity it is guaranteed that their corresponding files are deleted.
-Some functions have been replaced by faster/shorter variants.
-The size of the services has been reduced by introducing an photo admin service which handles all tasks related to adding photos to an album.
-Overall code consistency has been improved.
-Documentation has been improved.
-Code formatting has been improved.

### CODE REVIEW

File: module/Photo/src/Photo/Listener/AlbumDate.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+{
+
+    public function prePersist($eventArgs)
+    {
+        $entity = $eventArgs->getEntity();
+        if ($entity->getResourceId() === 'album') {
+            $this->albumPersisted($entity);
+        } elseif ($entity->getResourceId() === 'photo') {
+            $this->photoPersisted($entity);
+        }
```
Comment (kokx) [inline module/Photo/src/Photo/Listener/AlbumDate.php]: Is het hier niet beter om `instanceof` te gebruiken om te checken wat voor entity het precies is? Het hoeft namelijk niet zo te zijn dat elke entity de `getResourceId()` methode implementeert.

File: module/Photo/src/Photo/Listener/Remove.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        $this->sm = $sm;
+    }
+
+    public function preRemove($eventArgs)
+    {
+        $entity = $eventArgs->getEntity();
+        if ($entity->getResourceId() === 'album') {
+            $this->albumRemoved($entity);
+        } elseif ($entity->getResourceId() === 'photo') {
```
Comment (kokx) [inline module/Photo/src/Photo/Listener/Remove.php]: Idem. Gebruik `instanceof`.

File: module/Photo/src/Photo/Service/Admin.php [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+                $path,
+                $config['small_thumb_size']['width'],
+                $config['small_thumb_size']['height']
+            ));
+            $mapper = $this->getPhotoMapper();
+            /**
+             * TODO: optionally we could use a transactional query here to make it
+             * completely fail-safe in case something goes wrong when moving the
+             * photo in the storeUploadedPhoto function. However it's very unlikely
+             * anything will go wrong when moving the photo.
```
Comment (kokx) [inline module/Photo/src/Photo/Service/Admin.php]: Why not do so? Doctrine has pretty nice transactional capabilities, that even tie in perfectly with code.

http://doctrine-orm.readthedocs.org/en/latest/reference/transactions-and-concurrency.html

File: module/Photo/src/Photo/Service/Admin.php [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+        }
+    }
+
+    public function upload($files, $album)
+    {
+        $imageValidator = new \Zend\Validator\File\IsImage(
+            array('magicFile' => false)
+        );
+        $extensionValidator = new \Zend\Validator\File\Extension(
+            array('JPEG', 'JPG', 'JFIF', 'TIFF', 'RIF', 'GIF', 'BMP', 'PNG')
```
Comment (kokx) [inline module/Photo/src/Photo/Service/Admin.php]: Badass! BMP!

File: module/Photo/src/Photo/Service/Admin.php (hunk 3 above)
Comment (jszanto) [inline module/Photo/src/Photo/Service/Admin.php]: Inderdaad een goed idee als dat eindelijk eens gaat gebeuren. (dit stond al een jaar zo in de master volgens mij)

**Pull request merged**